{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a0e4e564",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f028453f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       0          1          2  3  4            5  \\\n",
      "0   chr2  175458592  175458593  T  A  SCLC32_R_NA   \n",
      "1   chr5   32303451   32303452  G  A  SCLC32_R_NA   \n",
      "2   chr3  186665823  186665824  G  A  SCLC32_R_NA   \n",
      "3   chr7  105162630  105162631  T  C  SCLC32_R_NA   \n",
      "4  chr20   33653254   33653255  C  T  SCLC32_R_NA   \n",
      "\n",
      "                                                   6  \n",
      "0  .;No;.;WIPF1:PPI(0.935);.;TFP,TFM,Enhancer;Yes...  \n",
      "1  .;No;.;MTMR12:PPI(0.142)REG(0.880);.;DHS,TFM,E...  \n",
      "2  .;No;.;ST6GAL1:PPI(0.361)REG(0.761);.;DHS,TFP,...  \n",
      "3  .;No;.;PUS7:PPI(0.707)REG(0.409);.;TFP,TFM;Yes...  \n",
      "4  .;No;.;TRPC4AP:PPI(0.540)REG(0.634);.;DHS,TFM;...  \n"
     ]
    }
   ],
   "source": [
    "SNP_candidates_df = pd.read_csv(\"/Users/myhoang/1_Griffithlab/SCLC/SCLC_funseq2/funseq2_khuranalab_DC3/SCLC_output/2023Feb09_recurrent_48samples/Candidates.Summary\", sep='\\t',low_memory=False,header=None)\n",
    "\n",
    "print(SNP_candidates_df.head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "77504f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of desired column names\n",
    "column_names = ['chr', 'start', 'end', 'ref', 'alt', 'sample', 'gerp;cds;variant.annotation.cds;network.hub;gene.under.negative.selection;ENCODE.annotated;hot.region;motif.analysis;sensitive;ultra.sensitive;ultra.conserved;target.gene[known_cancer_gene/TF_regulating_known_cancer_gene,differential_expressed_in_cancer,actionable_gene];coding.score;noncoding.score;recurrence;recurrence.cancer.dataset']\n",
    "\n",
    "# Set the column names for the DataFrame\n",
    "SNP_candidates_df.columns = column_names\n",
    "\n",
    "# Display the DataFrame with the new column names\n",
    "# print(SNP_candidates_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bdaade8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         chr      start        end ref alt        sample gerp cds  \\\n",
      "0       chr2  175458592  175458593   T   A   SCLC32_R_NA    .  No   \n",
      "1       chr5   32303451   32303452   G   A   SCLC32_R_NA    .  No   \n",
      "2       chr3  186665823  186665824   G   A   SCLC32_R_NA    .  No   \n",
      "3       chr7  105162630  105162631   T   C   SCLC32_R_NA    .  No   \n",
      "4      chr20   33653254   33653255   C   T   SCLC32_R_NA    .  No   \n",
      "...      ...        ...        ...  ..  ..           ...  ...  ..   \n",
      "77588   chr6   73014031   73014032   G   A  SCLC60_TN_LU    .  No   \n",
      "77589   chr6   73050581   73050582   G   T  SCLC60_TN_LU    .  No   \n",
      "77590   chr6   73088828   73088829   C   A  SCLC60_TN_LU    .  No   \n",
      "77591   chrX  154016432  154016433   C   T  SCLC60_TN_LU    .  No   \n",
      "77592   chrX  154036865  154036866   G   T  SCLC60_TN_LU    .  No   \n",
      "\n",
      "      variant.annotation.cds                   network.hub  ... hot.region  \\\n",
      "0                          .              WIPF1:PPI(0.935)  ...        Yes   \n",
      "1                          .   MTMR12:PPI(0.142)REG(0.880)  ...        Yes   \n",
      "2                          .  ST6GAL1:PPI(0.361)REG(0.761)  ...        Yes   \n",
      "3                          .     PUS7:PPI(0.707)REG(0.409)  ...        Yes   \n",
      "4                          .  TRPC4AP:PPI(0.540)REG(0.634)  ...        Yes   \n",
      "...                      ...                           ...  ...        ...   \n",
      "77588                      .              RIMS1:PPI(0.787)  ...          .   \n",
      "77589                      .              RIMS1:PPI(0.787)  ...          .   \n",
      "77590                      .              RIMS1:PPI(0.787)  ...          .   \n",
      "77591                      .    MPP1:PHOS(0.786)PPI(0.657)  ...          .   \n",
      "77592                      .    MPP1:PHOS(0.786)PPI(0.657)  ...          .   \n",
      "\n",
      "                                          motif.analysis sensitive  \\\n",
      "0      MOTIFBR:CHD2_FOS_MAX_MYC_NFKB1_PAX5_SP1_TAF1_T...         .   \n",
      "1                                            MOTIFBR:DHS         .   \n",
      "2                                            MOTIFBR:DHS         .   \n",
      "3      MOTIFBR:BHLHE40_CCNT2_E2F1_E2F4_E2F6_ELF1_ETS1...         .   \n",
      "4                                            MOTIFBR:DHS         .   \n",
      "...                                                  ...       ...   \n",
      "77588                                                  .         .   \n",
      "77589                                                  .         .   \n",
      "77590                                                  .         .   \n",
      "77591                                                  .         .   \n",
      "77592                                                  .         .   \n",
      "\n",
      "      ultra.sensitive ultra.conserved  \\\n",
      "0                   .               .   \n",
      "1                   .               .   \n",
      "2                   .               .   \n",
      "3                   .               .   \n",
      "4                   .               .   \n",
      "...               ...             ...   \n",
      "77588               .               .   \n",
      "77589               .               .   \n",
      "77590               .               .   \n",
      "77591               .               .   \n",
      "77592               .               .   \n",
      "\n",
      "      target.gene[known_cancer_gene/TF_regulating_known_cancer_gene,differential_expressed_in_cancer,actionable_gene]  \\\n",
      "0                                          WIPF1(Intron)                                                                \n",
      "1                                         MTMR12(Intron)                                                                \n",
      "2                                        ST6GAL1(Intron)                                                                \n",
      "3                                              PUS7(UTR)                                                                \n",
      "4                                        TRPC4AP(Intron)                                                                \n",
      "...                                                  ...                                                                \n",
      "77588                                      RIMS1(Intron)                                                                \n",
      "77589                                      RIMS1(Intron)                                                                \n",
      "77590                                      RIMS1(Intron)                                                                \n",
      "77591                                       MPP1(Intron)                                                                \n",
      "77592                                       MPP1(Intron)                                                                \n",
      "\n",
      "      coding.score   noncoding.score recurrence recurrence.cancer.dataset  \n",
      "0                .  3.59698430865953        Yes                Lung_Adeno  \n",
      "1                .  2.46772357202517          .                         .  \n",
      "2                .  2.25169503657219          .                         .  \n",
      "3                .  2.16571953646854          .                         .  \n",
      "4                .  2.09588623620306          .                         .  \n",
      "...            ...               ...        ...                       ...  \n",
      "77588            .  1.50326973506507        Yes                         .  \n",
      "77589            .  1.50326973506507        Yes                         .  \n",
      "77590            .  1.50326973506507        Yes                         .  \n",
      "77591            .  1.50164529133981        Yes                         .  \n",
      "77592            .  1.50164529133981        Yes                         .  \n",
      "\n",
      "[77593 rows x 22 columns]\n",
      "SNP_candidates_df shape (77593, 22)\n"
     ]
    }
   ],
   "source": [
    "SNP_candidates_df[['gerp', 'cds', 'variant.annotation.cds', 'network.hub', 'gene.under.negative.selection', 'ENCODE.annotated', 'hot.region', 'motif.analysis', 'sensitive', 'ultra.sensitive', 'ultra.conserved', 'target.gene[known_cancer_gene/TF_regulating_known_cancer_gene,differential_expressed_in_cancer,actionable_gene]', 'coding.score', 'noncoding.score', 'recurrence', 'recurrence.cancer.dataset']] = SNP_candidates_df['gerp;cds;variant.annotation.cds;network.hub;gene.under.negative.selection;ENCODE.annotated;hot.region;motif.analysis;sensitive;ultra.sensitive;ultra.conserved;target.gene[known_cancer_gene/TF_regulating_known_cancer_gene,differential_expressed_in_cancer,actionable_gene];coding.score;noncoding.score;recurrence;recurrence.cancer.dataset'].str.split(';', expand=True)\n",
    "\n",
    "SNP_candidates_df = SNP_candidates_df.drop(columns=['gerp;cds;variant.annotation.cds;network.hub;gene.under.negative.selection;ENCODE.annotated;hot.region;motif.analysis;sensitive;ultra.sensitive;ultra.conserved;target.gene[known_cancer_gene/TF_regulating_known_cancer_gene,differential_expressed_in_cancer,actionable_gene];coding.score;noncoding.score;recurrence;recurrence.cancer.dataset'])\n",
    "\n",
    "print(SNP_candidates_df)\n",
    "\n",
    "print('SNP_candidates_df shape',SNP_candidates_df.shape) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1930f524",
   "metadata": {},
   "source": [
    "### Data cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3f9ed630",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP_candidates_motifEvents_df shape before cleaning (28618, 22)\n",
      "SNP_candidates_motifEvents_df shape after removing MOTIFBR and MOTIFG both substring rows (28610, 22)\n"
     ]
    }
   ],
   "source": [
    "# DATA cleaning: \n",
    "\n",
    "# first, create SNP_candidates_motifEvents_df by subsetting SNP_candidates_df for columns which motif.analysis has either 'MOTIFBR' or 'MOTIFG' substring\n",
    "SNP_candidates_motifEvents_df = SNP_candidates_df[SNP_candidates_df['motif.analysis'].str.contains('MOTIFBR|MOTIFG', na=False)] \n",
    "\n",
    "print('SNP_candidates_motifEvents_df shape before cleaning',SNP_candidates_motifEvents_df.shape)\n",
    "\n",
    "# find the rows which contain both 'MOTIFBR' and 'MOTIFG' substrings in motif.analysis column\n",
    "SNP_candidates_motifBR_and_MOTIFG_df = SNP_candidates_df[SNP_candidates_df['motif.analysis'].str.contains('MOTIFBR', na=False) & SNP_candidates_df['motif.analysis'].str.contains('MOTIFG', na=False)]\n",
    "# print('SNP_candidates_motifBR_and_MOTIFG_df shape',SNP_candidates_motifBR_and_MOTIFG_df.shape)\n",
    "# print(SNP_candidates_motifBR_and_MOTIFG_df)\n",
    "\n",
    "# remove the rows which contain both 'MOTIFBR' and 'MOTIFG' substrings in motif.analysis column from SNP_candidates_motifEvents_df\n",
    "# this will leave a cleaned dataframe with rows that only contains either 'MOTIFBR' or 'MOTIFG' substring in motif.analysis column\n",
    "SNP_candidates_motifEvents_df = SNP_candidates_motifEvents_df[~(SNP_candidates_motifEvents_df.index.isin(SNP_candidates_motifBR_and_MOTIFG_df.index))]\n",
    "print('SNP_candidates_motifEvents_df shape after removing MOTIFBR and MOTIFG both substring rows',SNP_candidates_motifEvents_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "104f4a6c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{',', '_', ':'}\n"
     ]
    }
   ],
   "source": [
    "# in column 'motif.analysis', find all character which are not alphabetic or numeric\n",
    "# print those characters\n",
    "import re\n",
    "\n",
    "\n",
    "all_special_characters = set()\n",
    "for entry in SNP_candidates_motifEvents_df['motif.analysis']:\n",
    "    special_characters = re.findall(r'[^a-zA-Z0-9]', entry)\n",
    "    all_special_characters.update(special_characters)   \n",
    "\n",
    "print(all_special_characters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1401d47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Example of motif.analysis column which contains character , : MOTIFBR:CTCF,CTCF_HNF4A_HNF4G_RAD21\n"
     ]
    }
   ],
   "source": [
    "# print an example of motif.analysis column which contains character ','\n",
    "\n",
    "for entry in SNP_candidates_motifEvents_df['motif.analysis']:\n",
    "    if ',' in entry:\n",
    "        print('Example of motif.analysis column which contains character , :', entry)\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8692f7ac",
   "metadata": {},
   "source": [
    "##### NOTE\n",
    "\n",
    "* the motif in MOTIFBR are separated by either ',' or '_'\n",
    "\n",
    "* the motif in MOTIFG are always separated by ',' (some of the motif gained have underscore in its name .eg. ELF2_1,ETS_known4)\n",
    "\n",
    "* there are motif_analysis_values that has '\\_' in it (eg. \"DHS_ELK4_EP300_GATA1\\_\"  ~ notice there's an '_' after GATA1)\n",
    "\n",
    "\n",
    "##### SOLUTION\n",
    "Will need to handle MOTIFBR and MOTIFG separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c39db008",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add 2 new columns: motif_event and motif_analysis_values\n",
    "\n",
    "# Initialize empty columns\n",
    "SNP_candidates_motifEvents_df['motif_event'] = ''\n",
    "SNP_candidates_motifEvents_df['motif_analysis_values'] = ''\n",
    "\n",
    "# Iterate through rows and split values or keep original value based on delimiter presence\n",
    "for index, row in SNP_candidates_motifEvents_df.iterrows():\n",
    "    split_values = row['motif.analysis'].split(':')\n",
    "    if len(split_values) == 2:\n",
    "        SNP_candidates_motifEvents_df.at[index, 'motif_event'] = split_values[0]\n",
    "        SNP_candidates_motifEvents_df.at[index, 'motif_analysis_values'] = split_values[1]\n",
    "    else:\n",
    "        SNP_candidates_motifEvents_df.at[index, 'motif_event'] = row['motif.analysis']\n",
    "        SNP_candidates_motifEvents_df.at[index, 'motif_analysis_values'] = row['motif.analysis']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a462888d",
   "metadata": {},
   "source": [
    "### MOTIF GAINING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "37bbf864",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7, 3)\n"
     ]
    }
   ],
   "source": [
    "# Extract rows with motif_event 'MOTIFG'\n",
    "motifg_rows = SNP_candidates_motifEvents_df[SNP_candidates_motifEvents_df['motif_event'] == 'MOTIFG']\n",
    "motifg_rows = motifg_rows[['sample','motif_event','motif_analysis_values']]\n",
    "print(motifg_rows.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "015dac46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             sample motif_event motif_analysis_values\n",
      "6109    SCLC35_R_NA      MOTIFG     ELF2_1,ETS_known4\n",
      "8487    SCLC37_R_NA      MOTIFG         ZNF143_known1\n",
      "26123  SCLC41_TN_NA      MOTIFG     ELF2_1,ETS_known4\n",
      "31465   SCLC43_R_NA      MOTIFG     ELF2_1,ETS_known4\n",
      "33391  SCLC43_TN_NA      MOTIFG     ELF2_1,ETS_known4\n",
      "37683  SCLC45_TN_LU      MOTIFG           HDAC2_disc6\n",
      "59326   SCLC55_R_LU      MOTIFG               TBX20_5\n"
     ]
    }
   ],
   "source": [
    "print(motifg_rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "45fc850b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique motifs gained (MOTIFG): 5\n",
      "Unique motifs gained (MOTIFG): {'HDAC2_disc6', 'ZNF143_known1', 'ETS_known4', 'ELF2_1', 'TBX20_5'}\n"
     ]
    }
   ],
   "source": [
    "# find the set of unique motifs that are gained (MOTIFG) across all samples\n",
    "uniq_motifs_gain_set = set()\n",
    "for entry in motifg_rows['motif_analysis_values']:\n",
    "    motifs = entry.split(',')\n",
    "    uniq_motifs_gain_set.update(motifs) \n",
    "print('Number of unique motifs gained (MOTIFG):', len(uniq_motifs_gain_set))\n",
    "print('Unique motifs gained (MOTIFG):', uniq_motifs_gain_set)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d88d07c",
   "metadata": {},
   "source": [
    "### MOTIF BREAKING"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7fc428c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28603, 3)\n"
     ]
    }
   ],
   "source": [
    "# Extract rows with motif_event 'MOTIFBR'\n",
    "motifbr_rows = SNP_candidates_motifEvents_df[SNP_candidates_motifEvents_df['motif_event'] == 'MOTIFBR']\n",
    "motifbr_rows = motifbr_rows[['sample','motif_event','motif_analysis_values']]\n",
    "print(motifbr_rows.shape)\n",
    "\n",
    "# check if any motif_analysis_values for broken motif contains the items in uniq_motifs_gain_set as a substring\n",
    "for entry in motifbr_rows['motif_analysis_values']:\n",
    "    for motif in uniq_motifs_gain_set:\n",
    "        if motif in entry:\n",
    "            print('Found motif from gain set in broken motif analysis values:', motif, 'in entry:', entry)\n",
    "                     \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "caaeb5b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if at least 1 row contains 'ELF2_1' motif. retursn True if found, else False.\n",
    "motifbr_rows['motif_analysis_values'].str.contains('ELF2_1').any()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8efacc3",
   "metadata": {},
   "source": [
    "that shows that none of the motif in gain set is in broken motif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09f09d4c",
   "metadata": {},
   "source": [
    "### Now create 'motif_analysis_value_set'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e52e68b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creat column 'motif_analysis_value_set' in motifg_rows which contains set of motifs from column 'motif_analysis_values'\n",
    "motifg_rows['motif_analysis_value_set'] = motifg_rows['motif_analysis_values'].apply(lambda x: set(x.split(','))) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cce1ca3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create column 'motif_analysis_value_set; in motifbr_rows which contains set of motifs from column 'motif_analysis_values'\n",
    "# note that the value in motif_analysis_values column can be separated by either ',' or '_' or both\n",
    "# and there's also possibility of the motif being blank\n",
    "# so in case the set has item '', remove that item from the set\n",
    "motifbr_rows['motif_analysis_value_set'] = motifbr_rows['motif_analysis_values'].apply(lambda x: set(re.split(r'[,_]', x)))\n",
    "motifbr_rows['motif_analysis_value_set'] = motifbr_rows['motif_analysis_value_set'].apply(lambda x: x - {''})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "49814f6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>motif_event</th>\n",
       "      <th>motif_analysis_values</th>\n",
       "      <th>motif_analysis_value_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SCLC32_R_NA</td>\n",
       "      <td>MOTIFBR</td>\n",
       "      <td>CTCF,CTCF_HNF4A_HNF4G_RAD21</td>\n",
       "      <td>{HNF4A, CTCF, HNF4G, RAD21}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        sample motif_event        motif_analysis_values  \\\n",
       "8  SCLC32_R_NA     MOTIFBR  CTCF,CTCF_HNF4A_HNF4G_RAD21   \n",
       "\n",
       "      motif_analysis_value_set  \n",
       "8  {HNF4A, CTCF, HNF4G, RAD21}  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reality check: see how 'motif_analysis_value_set' for a row with both ',' and '_' looks like\n",
    "motifbr_rows[motifbr_rows['motif_analysis_values'] == 'CTCF,CTCF_HNF4A_HNF4G_RAD21']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e62715e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>motif_event</th>\n",
       "      <th>motif_analysis_values</th>\n",
       "      <th>motif_analysis_value_set</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14404</th>\n",
       "      <td>SCLC39_R_LI</td>\n",
       "      <td>MOTIFBR</td>\n",
       "      <td>DHS_ELK4_EP300_GATA1_,DHS_ELK4_EP300_GATA1_JUND</td>\n",
       "      <td>{GATA1, JUND, DHS, ELK4, EP300}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18792</th>\n",
       "      <td>SCLC39_TN_LI</td>\n",
       "      <td>MOTIFBR</td>\n",
       "      <td>DHS_ELK4_EP300_GATA1_,DHS_ELK4_EP300_GATA1_JUND</td>\n",
       "      <td>{GATA1, JUND, DHS, ELK4, EP300}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             sample motif_event  \\\n",
       "14404   SCLC39_R_LI     MOTIFBR   \n",
       "18792  SCLC39_TN_LI     MOTIFBR   \n",
       "\n",
       "                                 motif_analysis_values  \\\n",
       "14404  DHS_ELK4_EP300_GATA1_,DHS_ELK4_EP300_GATA1_JUND   \n",
       "18792  DHS_ELK4_EP300_GATA1_,DHS_ELK4_EP300_GATA1_JUND   \n",
       "\n",
       "              motif_analysis_value_set  \n",
       "14404  {GATA1, JUND, DHS, ELK4, EP300}  \n",
       "18792  {GATA1, JUND, DHS, ELK4, EP300}  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reality check 2: see how 'motif_analysis_value_set' for a row with blank motif_analysis_values looks like\n",
    "motifbr_rows[motifbr_rows['motif_analysis_values'].str.contains('DHS_ELK4_EP300_GATA1_')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "926529dd",
   "metadata": {},
   "source": [
    "### Create table of count of motif for each sample. \n",
    "each count is one snp\n",
    "a snp can be associated with multiple motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9aa052f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(28610, 4)\n"
     ]
    }
   ],
   "source": [
    "# first, combine the motifg_rows and motifbr_rows dataframes\n",
    "motifAll_rows = pd.concat([motifg_rows, motifbr_rows], ignore_index=True)\n",
    "print(motifAll_rows.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "938ac09f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(46, 2)\n"
     ]
    }
   ],
   "source": [
    "# create a list out of the set\n",
    "motifAll_rows['motif_analysis_value_list'] = motifAll_rows['motif_analysis_value_set'].apply(lambda x: list(x))\n",
    "\n",
    "# create motifAll_sample_df by grouping motifAll_rows by 'sample' and summing the 'motif_analysis_value_list' column\n",
    "motifAll_sample_df = motifAll_rows.groupby('sample')['motif_analysis_value_list'].sum().reset_index()\n",
    "\n",
    "print(motifAll_sample_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4e0028d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "\n",
    "motif_count_df = pd.DataFrame(motifAll_sample_df['motif_analysis_value_list'].apply(lambda x: Counter(x)).tolist()).fillna(0).astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "bcd04f84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        sample                          motif_analysis_value_list  MYC  TBP  \\\n",
      "0  SCLC32_R_NA  [MYC, TBP, NFKB1, SP1, PAX5, TAF1, FOS, WRNIP1...    4    4   \n",
      "1  SCLC33_R_LU  [JUND, IRF1, TBP, E2F4, RFX5, JUN, BRCA1, ATF3...   37   29   \n",
      "\n",
      "   NFKB1  SP1  PAX5  TAF1  FOS  WRNIP1  ...  ZNF143_known1  GTF3C2  ESRRA  \\\n",
      "0      3    2     2     6    3       1  ...              0       0      0   \n",
      "1     33   29    25    22   33       0  ...              0       0      0   \n",
      "\n",
      "   ZZZ3  BRF1  HDAC8  NR4A1  SREBF2  HDAC2_disc6  TBX20_5  \n",
      "0     0     0      0      0       0            0        0  \n",
      "1     0     0      0      0       0            0        0  \n",
      "\n",
      "[2 rows x 129 columns]\n"
     ]
    }
   ],
   "source": [
    "# join the motif_count_df with the motifAll_sample_df\n",
    "motifAll_sample_df = pd.concat([motifAll_sample_df, motif_count_df], axis=1)\n",
    "print(motifAll_sample_df.head(2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f62a8c16",
   "metadata": {},
   "source": [
    "### Perform differential analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d2fe5750",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample counts by type:\n",
      "Sample_type\n",
      "R     28\n",
      "TN    18\n",
      "Name: count, dtype: int64\n",
      "\n",
      "TN: 18, R: 28, Unknown: 0\n"
     ]
    }
   ],
   "source": [
    "# Add column 'Sample_type' to motifAll_sample_df based on sample names\n",
    "# if the sample name has substring 'TN' between the first and second underscore, then it's 'TN' sample type\n",
    "# if the sample name has substring 'R' between the first and second underscore, then it's 'R' sample type\n",
    "# else it's 'Unknown' sample type\n",
    "def determine_sample_type(sample_name):\n",
    "    parts = sample_name.split('_')\n",
    "    if len(parts) > 2:\n",
    "        if 'TN' in parts[1]:\n",
    "            return 'TN'\n",
    "        elif 'R' in parts[1]:\n",
    "            return 'R'\n",
    "    return 'Unknown'\n",
    "motifAll_sample_df['Sample_type'] = motifAll_sample_df['sample'].apply(determine_sample_type)\n",
    "\n",
    "# print the number of samples which are TN, R, and Unknown\n",
    "sample_counts = motifAll_sample_df['Sample_type'].value_counts()\n",
    "print(\"Sample counts by type:\")\n",
    "print(sample_counts)\n",
    "print(f\"\\nTN: {sample_counts.get('TN', 0)}, R: {sample_counts.get('R', 0)}, Unknown: {sample_counts.get('Unknown', 0)}\")\n",
    "\n",
    "# move 'Sample_type' columns to be the second column\n",
    "cols = motifAll_sample_df.columns.tolist()\n",
    "cols.insert(1, cols.pop(cols.index('Sample_type')))\n",
    "motifAll_sample_df = motifAll_sample_df[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "68283a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save data\n",
    "motifAll_sample_df_path = \"/Users/myhoang/1_Griffithlab/SCLC/SCLC_funseq2/funseq2_khuranalab_DC3/SCLC_output/2023Feb09_recurrent_48samples/data_interpretation_hg19/all_46samples/table/motif_sample_count_77kCandidates.tsv\"\n",
    "motifAll_sample_df.to_csv(motifAll_sample_df_path, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97c16a22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "127\n",
      "114\n",
      "['MYC', 'TBP', 'NFKB1', 'SP1', 'PAX5', 'TAF1', 'FOS', 'WRNIP1', 'YY1', 'CHD2']\n"
     ]
    }
   ],
   "source": [
    "# Filter motifs (>=20% non-zero) and run Mann–Whitney U (TN vs R)\n",
    "from scipy.stats import mannwhitneyu\n",
    "import numpy as np\n",
    "\n",
    "# Define the motif columns (as columns which are not base columns)\n",
    "base_cols = {\"sample\", \"Sample_type\", \"motif_analysis_value_list\"}\n",
    "motif_columns = [c for c in motifAll_sample_df.columns if c not in base_cols]\n",
    "# Keep only numeric motif columns\n",
    "motif_columns = [c for c in motif_columns if np.issubdtype(motifAll_sample_df[c].dtype, np.number)]\n",
    "# Drop empty motif name if present\n",
    "motif_columns = [c for c in motif_columns if c != \"\"]\n",
    "\n",
    "# Compute non-zero proportion across all samples\n",
    "# only keep motifs which more than 20% of the samples (both TN and R) are non zero\n",
    "nonzero_prop = (motifAll_sample_df[motif_columns] > 0).sum() / len(motifAll_sample_df)\n",
    "keep_motifs = nonzero_prop[nonzero_prop >= 0.20].index.tolist()\n",
    "\n",
    "print(len(motif_columns))\n",
    "print(len(keep_motifs))\n",
    "print(keep_motifs[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8d0e0b41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tests run on 114 motifs\n",
      "       motif  prop_nonzero  n_TN  n_R  median_TN  median_R  U_statistic  \\\n",
      "28      MAFK      0.956522    18   28       26.0      13.0        312.5   \n",
      "27      MAFF      0.956522    18   28       13.5       7.5        308.0   \n",
      "113   GTF3C2      0.326087    18   28        0.0       0.0        210.5   \n",
      "11       DHS      1.000000    18   28      476.5     360.0        298.0   \n",
      "70    ZBTB33      0.782609    18   28        2.5       2.0        296.0   \n",
      "15   BHLHE40      0.739130    18   28        1.0       1.0        294.5   \n",
      "63      NFE2      0.608696    18   28        1.0       1.0        293.5   \n",
      "91     PRDM1      0.608696    18   28        1.0       1.0        293.5   \n",
      "79     CTBP2      0.760870    18   28        2.5       1.0        291.0   \n",
      "96      BATF      0.869565    18   28        7.0       5.5        289.5   \n",
      "\n",
      "      p_value  \n",
      "28   0.176552  \n",
      "27   0.210165  \n",
      "113  0.261784  \n",
      "11   0.305793  \n",
      "70   0.321149  \n",
      "15   0.323623  \n",
      "63   0.331642  \n",
      "91   0.333090  \n",
      "79   0.377923  \n",
      "96   0.403050  \n"
     ]
    }
   ],
   "source": [
    "rows = []\n",
    "for motif in keep_motifs:\n",
    "    tn_vals = motifAll_sample_df.loc[motifAll_sample_df[\"Sample_type\"] == \"TN\", motif]\n",
    "    r_vals = motifAll_sample_df.loc[motifAll_sample_df[\"Sample_type\"] == \"R\", motif]\n",
    "\n",
    "    # Mann–Whitney U test (two-sided)\n",
    "    try:\n",
    "        U, p = mannwhitneyu(tn_vals, r_vals, alternative=\"two-sided\", method=\"auto\")\n",
    "    except TypeError:\n",
    "        U, p = mannwhitneyu(tn_vals, r_vals, alternative=\"two-sided\")\n",
    "\n",
    "    rows.append({\n",
    "        \"motif\": motif,\n",
    "        \"prop_nonzero\": float(nonzero_prop[motif]),\n",
    "        \"n_TN\": int((motifAll_sample_df[\"Sample_type\"] == \"TN\").sum()),\n",
    "        \"n_R\": int((motifAll_sample_df[\"Sample_type\"] == \"R\").sum()),\n",
    "        \"median_TN\": float(np.median(tn_vals)),\n",
    "        \"median_R\": float(np.median(r_vals)),\n",
    "        \"U_statistic\": float(U),\n",
    "        \"p_value\": float(p),\n",
    "    })\n",
    "\n",
    "mannwhitney_results = pd.DataFrame(rows).sort_values(\"p_value\")\n",
    "print(\"Tests run on\", len(mannwhitney_results), \"motifs\")\n",
    "print(mannwhitney_results.head(10))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "40e267ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       motif  prop_nonzero  n_TN  n_R  median_TN  median_R  U_statistic  \\\n",
      "28      MAFK      0.956522    18   28       26.0      13.0        312.5   \n",
      "27      MAFF      0.956522    18   28       13.5       7.5        308.0   \n",
      "113   GTF3C2      0.326087    18   28        0.0       0.0        210.5   \n",
      "11       DHS      1.000000    18   28      476.5     360.0        298.0   \n",
      "70    ZBTB33      0.782609    18   28        2.5       2.0        296.0   \n",
      "15   BHLHE40      0.739130    18   28        1.0       1.0        294.5   \n",
      "63      NFE2      0.608696    18   28        1.0       1.0        293.5   \n",
      "91     PRDM1      0.608696    18   28        1.0       1.0        293.5   \n",
      "79     CTBP2      0.760870    18   28        2.5       1.0        291.0   \n",
      "96      BATF      0.869565    18   28        7.0       5.5        289.5   \n",
      "\n",
      "      p_value  p_value_FDR  reject_null  \n",
      "28   0.176552          1.0        False  \n",
      "27   0.210165          1.0        False  \n",
      "113  0.261784          1.0        False  \n",
      "11   0.305793          1.0        False  \n",
      "70   0.321149          1.0        False  \n",
      "15   0.323623          1.0        False  \n",
      "63   0.331642          1.0        False  \n",
      "91   0.333090          1.0        False  \n",
      "79   0.377923          1.0        False  \n",
      "96   0.403050          1.0        False  \n"
     ]
    }
   ],
   "source": [
    "# Perform Benjamini-Hochberg (FDR) correction\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "reject, pvals_corrected, _, _ = multipletests(mannwhitney_results[\"p_value\"], method=\"fdr_bh\")\n",
    "mannwhitney_results[\"p_value_FDR\"] = pvals_corrected\n",
    "mannwhitney_results[\"reject_null\"] = reject\n",
    "print(mannwhitney_results.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "67674ecf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Significant motifs after FDR correction: 0\n",
      "Empty DataFrame\n",
      "Columns: [motif, prop_nonzero, n_TN, n_R, median_TN, median_R, U_statistic, p_value, p_value_FDR, reject_null]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# check if any motif has significant difference between TN and R after FDR correction\n",
    "significant_motifs = mannwhitney_results[mannwhitney_results[\"reject_null\"]]\n",
    "print(\"Significant motifs after FDR correction:\", significant_motifs.shape[0])\n",
    "print(significant_motifs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e90dbcad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save results\n",
    "out_path = \"/Users/myhoang/1_Griffithlab/SCLC/SCLC_funseq2/funseq2_khuranalab_DC3/SCLC_output/2023Feb09_recurrent_48samples/data_interpretation_hg19/all_46samples/table/motif_mannwhitney_TN_vs_R.tsv\"\n",
    "mannwhitney_results.to_csv(out_path, sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pvactools_MH_June2023_env_kernel",
   "language": "python",
   "name": "pvactools_mh_june2023_env_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
